Ability to add and update SSP Impl Req

rgauss opened the pull request:
- Created an `OscalSspImplReqMarshallerImpl`
- Updated `IterableAssemblyClassBinding` to define a list of
`secondaryRootObjects` that will 'trick' the deserializer to allow as
root objects
- Updated `BaseOscalObjectMarshallerLiboscalImpl` to be able to create
deserializer with `IterableAssemblyClassBinding`
- Updated `SspController` and tests

Closes #57
Closes #58

rgauss committed:
Ability to add and update SSP Impl Req

- Created an `OscalSspImplReqMarshallerImpl`
- Updated `IterableAssemblyClassBinding` to define a list of
`secondaryRootObjects` that will 'trick' the deserializer to allow as
root objects
- Updated `BaseOscalObjectMarshallerLiboscalImpl` to be able to create
deserializer with `IterableAssemblyClassBinding`
- Updated `SspController` and tests

laurelmay commented on oscal-data-repository-commons/src/main/java/com/easydynamics/oscal/data/marshalling/IterableAssemblyClassBinding.java:
```diff
@@ -4,13 +4,29 @@
+
+  static {
+    secondaryRootObjects = new HashedMap<>();
+    secondaryRootObjects.put(ImplementedRequirement.class,
+        new QName("implemented-requirement"));
```
So eventually, we'd also put something like:

```java
secondaryRootObjects.put(Control.class, new QName("control"));
secondaryRootObjects.put(Component.class, new QName("component"));
```

?

rgauss replied:
That's correct.

laurelmay commented on oscal-data-repository-commons/src/main/java/com/easydynamics/oscal/data/marshalling/IterableAssemblyClassBinding.java:
```diff
@@ -4,13 +4,29 @@
+   * Map of objects that are not declared as root in liboscal-java,
+   * but we want to treat them as such.
+   */
+  private static final Map<Class<?>, QName> secondaryRootObjects;
+
+  static {
+    secondaryRootObjects = new HashedMap<>();
+    secondaryRootObjects.put(ImplementedRequirement.class,
+        new QName("implemented-requirement"));
+  }
+
```
Because we only ever use the interface here, we never write to it after building, and we're Java 11+, we could potentially just use [`Map.ofEntries`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Map.html#ofEntries(java.util.Map.Entry...)). But if there's a specific reason to use `HashedMap`, this may not work for us.

I am also pretty biased against initializers.

```suggestion
  private static final Map<Class<?>, QName> secondaryRootObjects = Map.ofEntries(
    Map.entry(ImplementedRequirement.class, new QName("implemented-requirement"),
  );
```

rgauss replied:
Addressed.

laurelmay commented on oscal-data-repository-commons/src/main/java/com/easydynamics/oscal/data/marshalling/IterableAssemblyClassBinding.java:
```diff
@@ -48,4 +64,18 @@ public void writeRootItems(
+
+  @Override
+  public QName getRootXmlQName() {
+    if (secondaryRootObjects.containsKey(getBoundClass())) {
+      return secondaryRootObjects.get(getBoundClass());
+    }
+    return super.getRootXmlQName();
+  }
```
```suggestion
    return secondaryRootObjects.getOrDefault(getBoundClass(), super.getRootXmlQName());
  }
```

rgauss replied:
Addressed.

laurelmay commented on oscal-rest-service-app/src/main/java/com/easydynamics/oscalrestservice/api/SspController.java:
```diff
@@ -23,13 +35,17 @@
       OscalSspService sspService,
-      OscalObjectMarshaller<SystemSecurityPlan> marshaller
+      OscalObjectMarshaller<SystemSecurityPlan> marshaller,
+      OscalObjectMarshaller<ImplementedRequirement> oscalSspImplReqtMarshaller
```
To make sure I understand, we're eventually going to need a Marshaller for Components as well?

rgauss replied:
Potentially.  We may end up exposing a generic implementation that can take a class as an argument.

laurelmay commented on oscal-data-repository-commons/src/main/java/com/easydynamics/oscal/data/marshalling/IterableAssemblyClassBinding.java:
```diff
@@ -4,13 +4,29 @@
 import java.io.IOException;
+import java.util.Map;
+import javax.xml.namespace.QName;
+import org.apache.commons.collections4.map.HashedMap;
```
Just curious, why is this preferable to `java.util.HashMap`?

rgauss replied:
Errant auto-complete.

laurelmay commented on oscal-rest-service-app/src/main/java/com/easydynamics/oscalrestservice/api/SspController.java:
```diff
@@ -75,4 +91,122 @@ public ResponseEntity<StreamingResponseBody> put(
+
+    // Find existing ImplementedRequirement if exists and merge or add
+    ImplementedRequirement existingImplReq = null;
+    if (existingSsp.getControlImplementation() != null
+        && existingSsp.getControlImplementation().getImplementedRequirements() != null) {
```
Really missing the `?.` operator right now

laurelmay commented on oscal-rest-service-app/src/main/java/com/easydynamics/oscalrestservice/api/SspController.java:
```diff
@@ -75,4 +91,122 @@ public ResponseEntity<StreamingResponseBody> put(
+    if (existingSsp.getControlImplementation() != null
+        && existingSsp.getControlImplementation().getImplementedRequirements() != null) {
+      existingImplReq = existingSsp.getControlImplementation().getImplementedRequirements().stream()
+          .filter(implReq -> incomingImplReq.getUuid().equals(implReq.getUuid()))
+          .findAny()
```
Do we check for duplicate IDs anywhere? Here we will pick one (potentially at "random") and mutate that one if the `filter` has >1 elements. I suppose envisioning that a user PUT/POST(ed) a "bad" document initially and now we're going to update, should we 400/409/500, update one of them, or update all of them?

Here, as I read it, we arbitrarily update one. And as a user, I'd probably expect none or all to be modified.

I also don't know that we need to answer this today in this ticket; but I do think we may want to at least create a new ticket for it if we don't answer.

rgauss replied:
The code definitely assumes the existing data has been validated (unique IDs), and I believe the marshaling (via liboscal-java) called on save is configured to perform some validation though I don't see a uniqueness constraint for Implemented Requirements UUID in particular on quick look through liboscal-java's generated models.

Another ticket for a deeper investigation of validation in general would be great.

laurelmay replied:
I'll write that ticket later today!

laurelmay commented on oscal-rest-service-app/src/main/java/com/easydynamics/oscalrestservice/api/OscalControllerExceptionAdvice.java:
```diff
@@ -22,7 +22,7 @@ String oscalObjectNotFoundHandler(OscalObjectNotFoundException exception) {
   @ResponseBody
   @ExceptionHandler(OscalObjectConflictException.class)
-  @ResponseStatus(HttpStatus.BAD_REQUEST)
+  @ResponseStatus(HttpStatus.CONFLICT)
```
Technically, I think this needs a spec update

rgauss replied:
Yes, EasyDynamics/oscal-rest#50.

laurelmay reviewed (commented):
Okay so _love_ this. I started to take a look at doing this and got a little put off (overwhelmed) by the fact this was the first time we're really handling any sort of non-universal "inner" elements (because like we play with UUID and Metadata but everything root has one of those). So not only is this cool because I want _this_ feature but I think this is cool because it's a really good starting point for all the features to come which is really exciting.

Nothing here is concerning enough to me to block merging but I do want to ask a few questions to make sure I understand what's going on before I fully approve (and of course, while I was making comments anyway, I made fun of Java once or twice).

rgauss committed:
Improved secondaryRootObjects handling

Commit:
Only return the new implemented requirement

Per the specification, we must return the _new_ implemented requirement
object; not the entire SSP. We can add somewhat more generic versions of
the existing `makeXxxResponse` methods for this use case (and we can
leave the old implementations as more convenient wrappers for the root
object types).

laurelmay reviewed (changes requested):
See https://github.com/EasyDynamics/oscal-rest-service/pull/95 for issues around spec compliance. It was a little too cumbersome to add as suggestions directly in this PR.

hreineck commented on oscal-rest-service-app/src/main/java/com/easydynamics/oscalrestservice/api/SspController.java:
```diff
@@ -75,4 +91,122 @@ public ResponseEntity<StreamingResponseBody> put(
+          .orElse(null);
+    }
+    if (existingImplReq != null && isCreateOnly) {
+      throw new OscalObjectConflictException("Implented Requirement already exists");
```
```suggestion
      throw new OscalObjectConflictException("Implemented Requirement already exists");
```

rgauss replied:
Addressed

hreineck commented on oscal-rest-service-app/src/main/java/com/easydynamics/oscalrestservice/api/SspController.java:
```diff
@@ -75,4 +91,122 @@ public ResponseEntity<StreamingResponseBody> put(
+      @Parameter @PathVariable String id,
+      @Parameter @PathVariable String implementedRequirementId,
+      @RequestBody String json) {
+    return updateImplementedRequirement(id, implementedRequirementId, json, false);
```
Should the PUT only update existing impl reqs? Or should it also add a new one? Not sure if we ever specified.

laurelmay replied:
I feel like we talked about this a little bit at one point when we added PUT/PATCH somewhere else (but I can't remember at the moment). I don't remember a strong decision being reached one way or another.

_Theoretically_ these two _should_ be identical:

```
POST /system-security-plans/bb44ca2a-385b-4820-a3ba-b541b9569969/control-implementations/implemented-requirements

{
  "uuid: "762b1d87-1515-4253-a2ad-0e58dcc645a1"
  ...
}
```

```
PUT /system-security-plans/bb44ca2a-385b-4820-a3ba-b541b9569969/control-implementations/implemented-requirements/762b1d87-1515-4253-a2ad-0e58dcc645a1

{
  "uuid": "762b1d87-1515-4253-a2ad-0e58dcc645a1",
  ...
}
```

I think the difference here would be a little more clear if we _didnt_ require the `uuid` in the POST and PUT body (which I think I'd actually almost prefer for the spec). So then if a client had a preference for a UUID for a new resource for some reason (if you've already created a resource that references it or something or you're maybe doing some kind of migration), then you specify it (again, with POST and PUT being equivalent).

So skipping a few implementation details and abstracting a bit, I feel like as a user, I'd want the API/service to work a bit like this

```mermaid
graph LR
  Init([Make a request]) --> Pref{Preference on UUID}
  Pref -->|Yes| PUT
  Pref -->|No| POST
```

And

```mermaid
graph LR
  POST([POST]) --> UUID{Is UUID specified in body}
  UUID -->|Yes| Error([Error])
  UUID -->|No| Z[Create new Impl Req<br/>and generate UUID]
```

```mermaid
graph LR
  PUT([PUT]) --> UUID{Is UUID specified in body}
  UUID -->|No| Exist
  UUID -->|Yes| UUIDMatch
  UUIDMatch{Does the body UUID<br/>match path UUID}
  UUIDMatch -->|No| Error([Error])
  UUIDMatch -->|Yes| Exist
  Exist{Does a resource with the<br/>UUID already exist}
  Exist -->|Yes| Update[Update matching resource]
  Exist -->|No| Create[Create new<br/>Impl Req with given UUID]
```

rgauss replied:
@kylelaker, I'd agree with this.

Can you please create an issue to add a generic version of this (not specific to impl req) in the `oscal-rest` repo as 'implementation guidance' documentation?

hreineck commented on oscal-rest-service-app/src/main/java/com/easydynamics/oscalrestservice/api/SspController.java:
```diff
@@ -23,13 +35,17 @@
 @RestController
 public class SspController extends BaseOscalController<SystemSecurityPlan> {
+  private final Logger logger = LoggerFactory.getLogger(this.getClass());
+  private final OscalObjectMarshaller<ImplementedRequirement> oscalSspImplReqtMarshaller;
```
```suggestion
  private final OscalObjectMarshaller<ImplementedRequirement> oscalSspImplReqMarshaller;
```

Maybe change this to match the naming convention of "Impl Req" used elsewhere

rgauss replied:
Addressed

hreineck committed:
Separate POST and PUT Endpoint Implementations for SSP Impl Req

Separate the implementation of the POST endpoint (that adds a new Impl Req)
and the PUT endpoint (that updates existing Impl Reqs) into two functions.
This is done to match the API spec which specifies different parameters for
the two endpoints; the POST endpoint does not need the Impl Req ID.

hreineck committed:
Update Tests POST SSP Impl Req

Update the tests to accomodate the changes to the POST SSP Impl Req
endpoint's parameters

hreineck committed:
Fix Linter Issues

Resolve linter complaints on SSP Impl Req implementation and tests

hreineck commented on oscal-rest-service-app/src/main/java/com/easydynamics/oscalrestservice/api/SspController.java:
```diff
@@ -75,4 +91,122 @@ public ResponseEntity<StreamingResponseBody> put(
+   * @param implementedRequirementId the Implemented Requirement uuid
+   * @param json the SSP contents
+   */
+  @PostMapping(value = "/system-security-plans/{id}/control-implementation/"
+      + "implemented-requirements/{implementedRequirementId}",
```
Per the API spec, the POST request does not include the `implementedRequirementID` as a parameter.

Kind of annoying because the ID validation cannot happen in the marshaller - the functionality probably needs to be separated between the POST and PUT.

hreineck reviewed (changes requested):
Awesome! Works great in my local Docker deployment.
Only a couple notes.

hreineck commented:
Per my comment about the POST endpoint parameters, see the commits I just pushed for a potential way to separate out the POST and PUT endpoints. If we want a different way or to change the API spec I can revert

rgauss committed:
Merge pull request #95 from EasyDynamics/suggestion/update-ssp-impl-req

Only return the new implemented requirement

rgauss committed:
Spelling fixes

laurelmay reviewed (approved):
Thanks for your changes! This looks good as far as I can tell. We still need to see why the Viewer doesn't update after a PUT but right now I'm not sure anyone else has reproduced that behavior and it also feels more likely to be a viewer bug.

rgauss commented:
> Thanks for your changes! This looks good as far as I can tell. We still need to see why the Viewer doesn't update after a PUT but right now I'm not sure anyone else has reproduced that behavior and it also feels more likely to be a viewer bug.

@kylelaker, see EasyDynamics/oscal-react-library#396
